ambry-account/src/main/java/com/github/ambry/account/mysql/AccountDao.java

```diff
@@ -505,24 +515,27 @@ private void executeDeleteDatasetVersionStatement(PreparedStatement statement, i
    * @param containerId the id of the container.
    * @param datasetName the name of the dataset.
    * @param version the version value of the dataset.
-   * @param expirationTimeMs the expiration time in milliseconds.
+   * @param timeToLiveInSeconds The dataset version level ttl.
+   * @param creationTimeInMs the creation time of the dataset version.
    * @param dataset the {@link Dataset} including the metadata.
-   * @return
+   * @param datasetVersionTtlEnabled set to true if dataset version ttl want to override the dataset level default ttl.
+   * @return the updated expiration time.
    * @throws SQLException
    * @throws AccountServiceException
    */
-  private long executeUpdateDatasetVersionIfExpiredSqlStatement(PreparedStatement statement,
-      int accountId, int containerId, String datasetName, long version, long expirationTimeMs, Dataset dataset)
-      throws SQLException, AccountServiceException {
+  private long executeUpdateDatasetVersionIfExpiredSqlStatement(PreparedStatement statement, int accountId,
+      int containerId, String datasetName, long version, long timeToLiveInSeconds, long creationTimeInMs,
+      Dataset dataset, boolean datasetVersionTtlEnabled) throws SQLException, AccountServiceException {
     statement.setLong(1, version);
-    long updatedExpirationTimeMs = expirationTimeMs;
-    if (expirationTimeMs == Infinite_Time && dataset.getExpirationTimeMs() == Infinite_Time) {
+    long updatedExpirationTimeMs;
+    if (datasetVersionTtlEnabled || dataset.getRetentionTimeInSeconds() == null) {
+      updatedExpirationTimeMs = Utils.addSecondsToEpochTime(creationTimeInMs, timeToLiveInSeconds);
+    } else {
+      updatedExpirationTimeMs = Utils.addSecondsToEpochTime(creationTimeInMs, dataset.getRetentionTimeInSeconds());
+    }
+    if (updatedExpirationTimeMs == Infinite_Time) {
```

Comment:
Is that possible updatedExpirationTimesMs == Infitinite_time?
Looks like in either the if condition or the else condition, it'll be creationTimeInMs + xxxx.

Reply:
Inside addSecondsToEpochTime, if either creationTimeInMs or timeToLiveInSeconds == Infitinite_time, the updatedExpirationTimesMs will be Infitinite_time.